### DISCUSSION TITLE
Could haves

### DISCUSSION THREAD

**Discussion by jhpratt** [rust-marker/design#3] [Ideas]
Use this discussion for suggestions of "could haves" for a custom linter. Please limit suggestions to one per top-level comment to allow for easier upvoting/downvoting and replies.

**Comment by jhpratt** [comment]
Lints could be run in parallel.

**Reply by DevinR528** [reply]
This would make `ui` testing harder if the ordering wasn't deterministic... but I like this idea, I'm sure we could figure out ways to make it consistently order the output (by span for example).

**Reply by jhpratt** [reply]
Afaik that's why it hasn't been done in the compiler and clippy so far.  Collecting into a `Vec` and emitting by span seems plausible. Even if that's only done with a certain flag passed (for performance reasons).

**Reply by jhpratt** [reply]
Any objection to having this as a goal? We have the ability to easily test lints as a long-term goal, so as long as we ensure determinism these goals shouldn't conflict. Running in serial for testing is a trivial way to ensure this, though there's certainly other ways.

**Reply by flip1995** [reply]
No objections for a long term goal. I assume this will require some work on rustc before we can make this possible anyway.

**Reply by jhpratt** [reply]
Certainly for integration with rustc some work will be needed. My understanding is that we are starting this with no rustc integration, only moving to that point when necessary (using rustc crates, yes, but not _integrating_ with rustc proper).

**Comment by jhpratt** [comment]
Lints could be automatically downloaded via a registry.

**Reply by jhpratt** [reply]
Any objections to having this as a long-term goal? It could be opt-in, such as `auto-detect = true`, avoiding the need to explicitly specify lint packages.

**Comment by xFrednet** [comment]
Lints could be renamed by the end user.

This would be interesting if a user wants to
* replace a lint with a drop-in replacement that is named differently
* use one lint twice with different features enabled

**Reply by jhpratt** [reply]
I don't particularly care for this to be honest. This would effectively limit the ability of lint authors as well, as they'd have to consider potential breakage because of end users renaming things. Feature gates are being addressed elsewhere.

**Reply by DevinR528** [reply]
> replace a lint with a drop-in replacement that is named differently

If the lints are just crates could this be accomplished by just changing the dependency? Or do you mean something different?

**Reply by jhpratt** [reply]
That's a use case, correct? Not the actual functionality. If a user can rename a lint similar to imports (`use foo::bar as qux;` equivalent), then what would happen if a lint author creates a new lint with the name that's already used by the end user via renaming?

**Reply by jhpratt** [reply]
Thoughts on this as a whole?

**Reply by xFrednet** [reply]
Sorry, I didn't see your replies (GH notifications for discussions are sometimes not ideal). Anyway, I was thinking of an API like the one that Cargo currently provides for renaming crates while adding them as a dependency. This would mean renaming them lint at the position where they are discovered. (Discussed in #6)

In cargo the syntax is:
```toml
custom_name = { version = "1.0", features = ['derive'], package = "serde" }
#                                                       ^^^^^^^^^^^^^^^^^ The original package name
```

I don't want to define a syntax yet or anything, but at least discuss the possibility of renaming lints this way. Possible solutions for this are also highly dependent on how lints will be discovered in the end. The current plan for the start as agreed in #6 could support a syntax like this.

**Reply by jhpratt** [reply]
Ah, you meant renaming lint _packages_, not the lints themselves. I've no objection to that, as it would fall naturally from any registry, crates.io or otherwise. if lints end up being auto-discovered, nothing would stop the end user from manually adding in their own (which would implicitly include renaming).

**Reply by xFrednet** [reply]
At the start, I actually meant both, supporting the renaming of lints and packages. However, renaming just the package should cover my use case and would be sufficient. :upside_down_face:

**Reply by jhpratt** [reply]
Just checked and apparently while you _can_ `use clippy as foo;`, it's not possible to then use `foo::bar` for changing the lint severity. This honestly feels like a bug — either it should work or _both_ should error.

But with relation to us, I think allowing renaming in `Cargo.toml` is fine. Any objections?

**Comment by jplatte** [comment]
Meta question: Isn't it usually [Must have, Should have, Nice to have] as opposed to [Must have, Should have, Could have]?

**Reply by jhpratt** [reply]
No? https://en.m.wikipedia.org/wiki/MoSCoW_method
